[Linux] Improve backtracer ELF/DWARF parsing performance.

al45tair opened the pull request:
On Linux we use FileImageSource to read data from ELF images.  We were doing this in a fairly naive manner, which turns out to be a performance issue with larger statically linked binaries on Linux.

Change FileImageSource to use mmap() instead.

rdar://117590155

al45tair committed:
[Linux] Improve backtracer ELF/DWARF parsing performance.

On Linux we use FileImageSource to read data from ELF images.  We were
doing this in a fairly naive manner, which turns out to be a performance
issue with larger statically linked binaries on Linux.

Change FileImageSource to use mmap() instead.

rdar://117590155

al45tair committed:
[Linux] Don't keep the file descriptor for FileImageSource.

We don't really need it after calling `mmap()`, and not keeping it around
means it's easy to make sure it gets closed properly in all cases.

rdar://117590155

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please test

tbkka commented:
I'm not convinced this meets the bar for 5.9.   How big of a performance issue is it?

stephentyrone commented:
"4x speedup on both x86 and Apple Silicon."

al45tair commented:
I think @tomerd is probably keen on it going into a 5.9 Linux-only update (it's approximately a 4x speed-up on parsing DWARF, which appears to matter for some of the Swift-on-Server folks who are statically linking large amounts of code into their binaries).

stephentyrone commented:
Is this already on main/5.10?

al45tair commented:
I haven't merged the PR on main yet, and I hadn't written the CCC thingummy yet either (I've been pondering what to do with another performance improvement PR — I think I'm going to rebase it on this set because it conflicts).

The original is here: https://github.com/apple/swift/pull/69508. The 5.10 one is here: https://github.com/apple/swift/pull/69603.

al45tair commented:
**Explanation**: On Linux, we read data from ELF files using `FileImageSource`, which implements the `ImageSource` and `MemoryReader` protocols to allow code to fetch data from the image easily. The existing implementation is naïve and relies solely on the buffer cache in the OS to provide reasonable performance. Changing the implementation to use `mmap()` instead yields an approximately 4x speed-up. Some of the server-side Swift developers churn out statically linked single binaries in the hundreds of MB range where simply parsing the DWARF data from the image takes significant numbers of seconds, hence the interest in this for a 5.9 update.
**Original PR**: #69508
**Reviewed by**: @mikeash
**Resolves**: rdar://117590155
**Tests**: The existing backtracing tests exercise this code.

al45tair committed:
[Linux] Improve the bounds checks.

It was pointed out that there was a more Swift-y way to write the
bounds checks for the `fetch<T>` implementation, so do that instead.

rdar://117590155

al45tair commented:
@swift-ci Please test

tomerd commented:
yes, I would like to see this on the next Linux/Windows 5.9 dot release, as at the moment this prevents backtraces in environments like kubernetes where the application is killed after 30s of inactivity (it takes more than 30s to print the backtraces without this fix)
